### ISSUE TITLE
[objc] "Make the word before the argument describe the argument" rule

### ISSUE THREAD

**Issue by VincentDondain** [mono/Embeddinator-4000#193]
Here is the rule from [Apple](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html#//apple_ref/doc/uid/20001282-BCIGIJJF):

-----
Make the word before the argument describe the argument.
- (id)viewWithTag:(NSInteger)aTag;
**Right.**
- (id)taggedView:(int)aTag;
**Wrong.**
-----

Given that, https://github.com/mono/Embeddinator-4000/blob/objc/tests/managed/methods.cs#L93-L94 don't respect this convention calling for a `With...` to describe the argument (even though we won't properly describe the argument) and would look weird to ObjC devs.

**Comment by spouliot** [comment]
Apple's guideline is for humans and sadly computer are not really good at describing (and we can't turn this into an AI research project). 

`With` is already used with `Init` since it's always (or near enough) the best choice. That's less true with methods, e.g.

* `drawInRect:` is not the same as `drawWithRect`
* `drawAtPoint:`
* `performSegueForRow:`

I'm not sure we can create code to always pick the best (it would need to understand English and the context) but there's likely some common patterns we could hard code. What we need is a list of the common patterns.

**Comment by VincentDondain** [comment]
Mmh, knowing that we wouldn't turn that into an AI research project and use cool algo + a mega neural network, I only raised the things in this doc I _thought_ were generic enough.

Isn't the `With` generic enough to cover `In`, `At` and `For`? It's not as precise but that's not my point.

```
- (int)doItInt32:(int) i;
- (int)doItString:(NSString *) s;
- (int)doItI:(int) i j:(int) j;
```

Those are super weird.

The problem is not having a separation between the "main method name" section (if that makes any sense) and the "first argument description".

**Comment by spouliot** [comment]
Heh, naming is hard :-) 

**The problem is not having a separation between the "main method name" section (if that makes any sense) and the "first argument description".** is a good recap of the issue. It's basically always present in ObjC while it's not in .NET API.

> Isn't the With generic enough to cover In, At and For? 

I don't think `With` is a drop-in replacement for the others, it can change the meaning of the action. 

E.g. are you going on vacation **with** the children or **for** the children ? maybe they just need a break, not a vacation ;-) but they could be angry if you don't bring them.

It's why `With` is, right now, only used with `Init` - I'm still hoping for a miracle word 😇 

Grep'ing around tonight I've seen `Using`... and it does not sound too bad with `- (int)doItUsingInt32:(int) i;` but maybe it's really bad in other places.

> Those are super weird.

TBH several of the names I added in our tests were because I knew they would pose naming issues 🙊 There's not many Apple API that starts with `Do` 🙈 

Finally we always have a (long term) solution to let people rename the API (e.g. using attributes) to make them feel more native to their platform.

**Comment by VincentDondain** [comment]
I'm okay with an other magical word, using seems fine too.

I actually thought about `With` exactly like you're thinking of `Using`. As in "main method name" **with** / **using** the first argument. It's myMehod **with** an argument.

*If I understand your point about going on vacations **with** the children or **for** the children but even if you're going **for** them you're still going **with** them :P*

So maybe what you don't like is the fact that **with** in ObjC is used elsewhere with a precise meaning and that for our generated code we should use an other key word to avoid confusing people too much.

That's fine by me.

**Comment by spouliot** [comment]
yeah, I'm not sure if it's worse to use `With` (as it won't be correct in several cases) or to skip the _middle_ preposition. Ideally a more complex rule (than just use it for `Init`) could solve this... I'm just looking for it :)

**Comment by VincentDondain** [comment]
Skipping the middle preposition is worse IMHO, as fusion the argument name with the method name can lead to incomprehensible method name (because you can't really parse the main method name) and can also transform the intent of the method.

It's a problem if your main method name is using a verb right? Because you're gonna concatenate the verb and the first argument name (in some cases) and that would mean something different.

-----

Examples of random code I found in md-addins to show how weird it is without a separation keyword:

```
public void OnInitializeFromTemplate (ProjectCreateInformation projectCreateInfo, XmlElement template) { }
```

which becomes (I cheated on the types):

```
- (void)onInitializeFromTemplateProjectCreateInfo:(MyLib_ProjectCreateInformation *) projectCreateInfo template:(MyLib_XmlElement *) template;
```

In ObjC here, I'm just confused is the main method `onInitializeFromTemplate`, `onInitializeFromTemplateProject` or are we initializing from a `TemplateProjectCreateInfo` :P

-----

```
public void AddDefaultWatchAppConfigurations (Project watchExtensionProject) { }
```

becomes:

```
- (void)addDefaultWatchAppConfigurationsWatchExtensionProject:(MyLib_Project *) watchExtensionProject;
```

-----

So yes I think not having the middle preposition is worse, it should make most user code unreadable in ObjC (:

**Comment by VincentDondain** [comment]
Also while we're on the subject, the simplest solution is still to not concatenate the first argument to the method's name if possible.

If you don't have any overloads it's likely safer (produces better code) to avoid using the first argument or type.

**Comment by timrisi** [comment]
While perhaps the simplest solution, it ALSO breaks the method naming conventions from apple if you drop the argument from the methods name

**Comment by VincentDondain** [comment]
You are right, I forgot to remove my last comment, I was looking at cases that were misleading and figured out afterward (:

The first argument **is always used** in the main method's name.

In some cases like Sebastien pointed out it's not with a middle preposition like:

```
- (BOOL)openFile:(NSString *)fullPath ...;
```

But even if not perfect it wouldn't be that bad/wrong with `With` and the argument name.

```
- (BOOL)openWithFullPath:(NSString *)fullPath ...;
```
